### PR TITLE
[Not for merge] Madam oom

### PR DESCRIPTION
**Pull request by csukuangfj**
Put it here for discussion. Not ready for merge.

It contains
- madam optimizer from Dan
- Handle OOM when using a larger `max-duration` (e.g., from 200 to 350)

### CODE REVIEW

File: egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+):
+    try:
+        return compute_loss_impl(
+            params=params,
+            model=model,
+            batch=batch,
+            graph_compiler=graph_compiler,
+            is_training=is_training,
+        )
+    except RuntimeError as ex:
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py:363]: This is where OOM is handled.  I am testing it. Not sure whether it works or not.

Comment (pzelasko) [reply egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py:363]: Not sure if it's helpful -- there is a repo that has some utilties for finding the optimal batch size in PyTorch and it has some code to deal with OOM. Maybe there is something useful that can be borrowed https://github.com/BlackHC/toma

File: egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py [hunk 2]
```diff
@@ -0,0 +1,767 @@
...
+#  from transformer import Noam
+from madam import Moam
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.nn.utils import clip_grad_norm_
+from torch.utils.tensorboard import SummaryWriter
+
+from icefall.bpe_graph_compiler import BpeCtcTrainingGraphCompiler
+from icefall.checkpoint import load_checkpoint
+from icefall.checkpoint import save_checkpoint as save_checkpoint_impl
+from icefall.dataset.librispeech import LibriSpeechAsrDataModule
```
Comment (pzelasko) [inline egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py:28]: I strongly suggest making this class local too (e.g. in a local `data.py` file) -- in the current repo layout, it will make it much easier to experiment with different types of data setups and augmentations.

File: egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py (hunk 1 above)
Comment (danpovey) [reply egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py:381]: Hmm, after looking into the DDP implementation, it looks to me like the backward gradient reduction is not done in one pass after the backward is completed on individual machines, but is done during the .backward() of the model.  So I think catching errors that happen during .backward() is not going to be possible, and likewise for errors that happen during the top-level model forward() function, because ddp seems to have sync points there.  Errors in the forward for CTC or transformer decoder may be catchable though.

Comment (danpovey) [reply egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py:381]: ... but we can't do a new top-level model forward on new data.  Might be possible to re-do a CTC or transformer forward with a subset of the minibatch, since its backward pass would be structurally similar to the one on the entire minibatch.

File: egs/librispeech/ASR/conformer_ctc_madam_no_warmup/transformer.py [hunk 3]
```diff
@@ -0,0 +1,993 @@
...
+
+        Returns:
+          A scalar tensor containing the loss without normalization.
+        """
+        assert x.size(2) == self.size
+        #  batch_size = x.size(0)
+        x = x.view(-1, self.size)
+        target = target.view(-1)
+        with torch.no_grad():
+            true_dist = x.clone()
```
Comment (danpovey) [inline egs/librispeech/ASR/conformer_ctc_madam_no_warmup/transformer.py:838]: instead of x.clone(), torch.empty_like(x) would be more approprate..

File: egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py (hunk 1 above)
Comment (csukuangfj) [reply egs/librispeech/ASR/conformer_ctc_madam_no_warmup/train.py:381]: The current code is able to catch CUDA OOM exceptions during `model.forward()` when trained with a single GPU,
but it hangs when trained with two GPUs using DDP. I am looking into where it hangs.

https://github.com/pytorch/pytorch/issues/18853#issuecomment-583779161
says it is possible to catch CUDA OOM exceptions also for DDP training with a single GPU.

---

We are currently catching exceptions only for the forward pass, i..e, `model.forward`, `get_tot_scores`, `attetion decoder`.

---

> So I think catching errors that happen during .backward() is not going to be possible,

Agreed.

File: egs/librispeech/ASR/conformer_ctc_madam_no_warmup/conformer.py [hunk 4]
```diff
@@ -869,7 +869,10 @@ def __init__(
             groups=channels,
             bias=bias,
         )
-        self.norm = nn.BatchNorm1d(channels)
+        # NOTE(fangjun): The process hangs when using DDP
+        # if we try to recover from CUDA OOM, so we disable
+        # batchnorm layer here.
+        #  self.norm = nn.BatchNorm1d(channels)
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conformer_ctc_madam_no_warmup/conformer.py:875]: After disabling batch norm, training with DDP can now recover from OOM in `model.forward()` as expected.

**Comment by csukuangfj** [comment]
I will port OOM handling to LF-MMI training as well.

File: egs/librispeech/ASR/conformer_ctc_madam_no_warmup/conformer.py (hunk 4 above)
Comment (danpovey) [reply egs/librispeech/ASR/conformer_ctc_madam_no_warmup/conformer.py:875]: Mm.  Hopefully with the madam optimizer, the training will still be stable without the batchnorm.  We'll have to see.  Obviously would have to compare the performance after this change.
